barman/cloud_providers/google_cloud_storage.py

```diff
@@ -333,3 +337,282 @@ def _delete_objects_batch(self, paths):
+        # so would break backup to GCS for anyone unable to install
+        # google-cloud-compute (which includes anyone using python 2.7).
+        compute = import_google_cloud_compute()
+
+        self.client = compute.SnapshotsClient()
+        self.disks_client = compute.DisksClient()
+        self.instances_client = compute.InstancesClient()
+
+    def _get_instance_metadata(self, instance_name, zone):
+        """
+        Retrieve the metadata for the named instance in the specified zone.
+
+        :rtype: google.cloud.compute_v1.types.Instance
+        :return: An object representing the compute instance.
+        """
+        try:
+            return self.instances_client.get(
+                instance=instance_name,
+                zone=zone,
+                project=self.project,
+            )
+        except NotFound:
+            raise SnapshotBackupException(
+                "Cannot find instance with name %s in zone %s for project %s"
+                % (instance_name, zone, self.project)
+            )
+
+    def _get_disk_metadata(self, disk_name, zone):
+        """
+        Retrieve the metadata for the named disk in the specified zone.
+
+        :rtype: google.cloud.compute_v1.types.Disk
+        :return: An object representing the disk.
+        """
+        try:
+            return self.disks_client.get(
+                disk=disk_name, zone=zone, project=self.project
+            )
+        except NotFound:
+            raise SnapshotBackupException(
+                "Cannot find disk with name %s in zone %s for project %s"
+                % (disk_name, zone, self.project)
+            )
+
+    def take_snapshot(self, backup_info, disk_zone, disk_name):
+        """
+        Take a snapshot of a persistent disk in GCP.
+
+        :param barman.infofile.LocalBackupInfo backup_info: Backup information.
+        :param str disk_zone: The zone in which the disk resides.
+        :param str disk_name: The name of the source disk for the snapshot.
+        :rtype: str
+        :return: The name used to reference the snapshot with GCP.
+        """
+        snapshot_name = "%s-%s-%s" % (
+            backup_info.server_name.lower(),
+            backup_info.backup_id.lower(),
+            disk_name,
+        )
+        _logger.info("Taking snapshot '%s' of disk '%s'", snapshot_name, disk_name)
+        resp = self.client.insert(
+            {
+                "project": self.project,
+                "snapshot_resource": {
+                    "name": snapshot_name,
+                    "source_disk": "projects/%s/zones/%s/disks/%s"
+                    % (
+                        self.project,
+                        disk_zone,
+                        disk_name,
+                    ),
+                },
+            }
+        )
+
+        _logger.info("Waiting for snapshot '%s' completion", snapshot_name)
+        resp.result()
+
+        if resp.done():
```

Comment:
Checking `.done()` to determine success is ok but the [examples given in the GCP docs](https://cloud.google.com/compute/docs/disks/create-snapshots#create_zonal_snapshot) use the presence of the `error_code` field to determine success status.

Additionally, the examples check the `.warning` field and log anything present, which seems like a practice worth following here.

Reply:
This was updated so we use error_code to check for success and log any warnings.